File: samples_C++/ex_addWeighted.cpp
#include <opencv2/opencv.hpp>

int main(int agrc,  char* argv[]){
    std::string name1 = "../data/lena.png";

    cv::Mat img1=cv::imread(name1, 0);
    cv::Mat img2;
    cv::flip(img1, img2, 1);
    cv::Mat img3;
    
    int64 time0 = cv::getTickCount();
    cv::addWeighted(img2, 0.5, img1, -0.5, 128, img3);
    int64 time1 = cv::getTickCount();

	cv::imwrite("addWeighted.png", img3);

	cv::Size size=cv::Size(img1.cols, img1.rows);
	cv::Mat img4 = cv::Mat::zeros(size, CV_8UC1);
	
    int64 time2 = cv::getTickCount();
	for(int y=0; y< img1.rows; y++){
		for(int x=0; x< img1.cols; x++){
			img4.at<uchar>(y,x) = 0.5*img2.at<uchar>(y,x) - 0.5* img1.at<uchar>(y,x) +128;
		}
	}
    int64 time3 = cv::getTickCount();
	cv::imwrite("addWeighted_loop.png", img4);
	
    std::cout << (time1-time0)/cv::getTickFrequency() << std::endl;
    std::cout << (time3-time2)/cv::getTickFrequency() << std::endl;

}

File: samples_Python/ex_addWeighted.py
import cv2
import numpy as np
import pylab as plt

img1 = cv2.imread("../data/lena.png", 0)
img2 = img1[:, ::-1]+0

t1 = cv2.getTickCount()
img3 = cv2.addWeighted(img2, 0.5, img1, -0.5, 128)
t2 = cv2.getTickCount()

cv2.imwrite("img3.png", img3)
t3 = cv2.getTickCount()
img4 = 0.5*img2-0.5*img1 +128
t4 = cv2.getTickCount()
cv2.imwrite("img4.png", img4)

t5 = cv2.getTickCount()
img5 = (np.array(img2, dtype=np.int16) -np.array(img1,dtype=np.int16))/2 +128
t6 = cv2.getTickCount()
cv2.imwrite("img5.png", img5)

print (t2-t1)/cv2.getTickFrequency()
print (t4-t3)/cv2.getTickFrequency()
print (t6-t5)/cv2.getTickFrequency()


plt.figure(1)
plt.clf()
plt.subplot(2, 3, 1)
plt.imshow(img3)
plt.colorbar()
plt.subplot(2, 3, 2)
plt.imshow(img4)
plt.colorbar()
plt.subplot(2, 3, 3)
plt.imshow(img5)
plt.colorbar()
plt.subplot(2, 3, 5)
plt.imshow(img4-img3)
plt.colorbar()
plt.subplot(2, 3, 6)
plt.imshow(img5-img3)
plt.colorbar()
plt.show()
